packages/@lwc/template-compiler/src/codegen/expression.ts

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { walk } from 'estree-walker';
+import * as t from '../shared/estree';
+import { ComplexExpression } from '../shared/types';
+import { TEMPLATE_PARAMS } from '../shared/constants';
+import type CodeGen from './codegen';
+
+type VariableName = string;
+type VariableShadowingMultiplicity = number;
+type VariableNames = Set<string>;
+
+/**
+ * Bind the passed expression to the component instance. It applies the following
+ * transformation to the expression:
+ *    {value} --> {$cmp.value}
+ *    {value[index]} --> {$cmp.value[$cmp.index]}
+ *    {foo ?? bar} --> {$cmp.foo ?? $cmp.bar}
+ *    {foo?.bar} --> {$cmp.foo?.bar}
+ *
+ * However, parameter variables are not be transformed in this way. For example,
+ * the following transformations do not happen:
+ *    {(foo) => foo && bar} -> {(foo) => $cmp.foo && $cmp.bar}
+ *    {(foo) => foo && bar} -> {($cmp.foo) => foo && $cmp.bar}
+ *    {(foo) => foo && bar} -> {($cmp.foo) => $cmp.foo && $cmp.bar}
+ *
+ * Instead, the scopes are respected:
+ *    {(foo) => foo && $cmp.bar}
+ *
+ * Similar checks occur for local identifiers introduced via for:each or similar.
+ */
+export function bindComplexExpression(
+    expression: ComplexExpression,
+    codeGen: CodeGen
+): t.Expression {
+    const expressionScopes = new ExpressionScopes();
+    walk(expression, {
+        enter(node, _parent) {
+            if (t.isArrowFunctionExpression(node)) {
```

Comment:
Might be worth mentioning in a comment that function expressions are not allowed, so we only need to check for arrow functions.

Reply:
Added a comment.